### PR TITLE
[FLINK-18974][docs-zh]Translate the 'User-Defined Functions' page of "Application Development's DataStream API" into Chinese

### PR DESCRIPTION
**Pull request by zhuxiaoshang**
## What is the purpose of the change

[FLINK-18974][docs-zh]Translate the 'User-Defined Functions' page of "Application Development's DataStream API" into Chinese

## Brief change log
  - Translate the 'User-Defined Functions' page of "Application Development's DataStream API" into Chinese

## Verifying this change

This change is a trivial rework / code cleanup without any test coverage.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (no)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: ( no)
  - The serializers: (no)
  - The runtime per-record code paths (performance sensitive): (no)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn/Mesos, ZooKeeper: (no)
  - The S3 file system connector: (no)

## Documentation

  - Does this pull request introduce a new feature? (no)
  - If yes, how is the feature documented? (no)

### CODE REVIEW

File: docs/dev/user_defined_functions.zh.md [hunk 1]
```diff
@@ -147,95 +142,75 @@ data.map (new RichMapFunction[String, Int] {
...
-creating and finalizing local state, accessing broadcast variables (see
-[Broadcast Variables]({{ site.baseurl }}/dev/batch/index.html#broadcast-variables)), and for accessing runtime
-information such as accumulators and counters (see
-[Accumulators and Counters](#accumulators--counters)), and information
-on iterations (see [Iterations]({{ site.baseurl }}/dev/batch/iterations.html)).
+除了用户自定义的功能（map，reduce 等），富函数还提供了四个方法：`open`、`close`、`getRuntimeContext` 和
+`setRuntimeContext`。这些对于参数化功能很有用
+(参阅 [给函数传递参数]({{ site.baseurl }}/zh/dev/batch/index.html#passing-parameters-to-functions))，
+创建和最终确定本地状态，访问广播变量(参阅
+[广播变量]({{ site.baseurl }}/zh/dev/batch/index.html#broadcast-variables))，以及访问运行时信息，例如累加器和计数器(参阅
```
Comment (RocMarshal) [inline docs/dev/user_defined_functions.zh.md:149]: The link-tags, liked as  `({{ site.baseurl }}/zh/dev/batch/index.html#broadcast-variables))` need to be improved by [Prefer format of link-tags](http://apache-flink-mailing-list-archive.1008284.n3.nabble.com/Reminder-Prefer-link-tag-in-documentation-td42362.html).

File: docs/dev/user_defined_functions.zh.md [hunk 2]
```diff
@@ -147,95 +142,75 @@ data.map (new RichMapFunction[String, Int] {
...
+[累加器和计数器](#累加器和计数器))，以及迭代器的相关信息(参阅 [迭代器]({{ site.baseurl }}/zh/dev/batch/iterations.html))。
```
Comment (RocMarshal) [inline docs/dev/user_defined_functions.zh.md:150]: The link tags in current page , liked as `[累加器和计数器](#累加器和计数器)`  need to be corrected  after the change which could be made by   the item `4. 标题锚点链接` of [Flink Translation Specifications](https://cwiki.apache.org/confluence/display/FLINK/Flink+Translation+Specifications)

File: docs/dev/user_defined_functions.zh.md [hunk 3]
```diff
@@ -147,95 +142,75 @@ data.map (new RichMapFunction[String, Int] {
...
-  and {% gh_link /flink-core/src/main/java/org/apache/flink/api/common/accumulators/DoubleCounter.java "__DoubleCounter__" %}:
-  See below for an example using a counter.
-- {% gh_link /flink-core/src/main/java/org/apache/flink/api/common/accumulators/Histogram.java "__Histogram__" %}:
-  A histogram implementation for a discrete number of bins. Internally it is just a map from Integer
-  to Integer. You can use this to compute distributions of values, e.g. the distribution of
-  words-per-line for a word count program.
+  和 {% gh_link /flink-core/src/main/java/org/apache/flink/api/common/accumulators/DoubleCounter.java "__DoubleCounter__" %}:
+  有关使用计数器的示例，请参见下文。
+- {% gh_link /flink-core/src/main/java/org/apache/flink/api/common/accumulators/Histogram.java "__直方图__" %}:
+  离散数量的柱状直方图实现。 在内部，它只是整形到整形的映射。你可以使用它来计算值的分布，例如，单词计数程序的每行单词的分布情况。
```
Comment (RocMarshal) [inline docs/dev/user_defined_functions.zh.md:171]:
```suggestion
  离散数量的柱状直方图实现。在内部，它只是整形到整形的映射。你可以使用它来计算值的分布，例如，单词计数程序的每行单词的分布情况。
```

File: docs/dev/user_defined_functions.zh.md [hunk 4]
```diff
@@ -147,95 +142,75 @@ data.map (new RichMapFunction[String, Int] {
...
 
-__How to use accumulators:__
+__如何使用累加器:__
 
-First you have to create an accumulator object (here a counter) in the user-defined transformation
-function where you want to use it.
+首先，你要在需要使用累加器的用户自定义的转换函数中创建一个累加器对象(此处是计数器) 。
```
Comment (RocMarshal) [inline docs/dev/user_defined_functions.zh.md:175]:
```suggestion
首先，你要在需要使用累加器的用户自定义的转换函数中创建一个累加器对象（此处是计数器）。
```

File: docs/dev/user_defined_functions.zh.md [hunk 5]
```diff
@@ -147,95 +142,75 @@ data.map (new RichMapFunction[String, Int] {
...
-  A histogram implementation for a discrete number of bins. Internally it is just a map from Integer
-  to Integer. You can use this to compute distributions of values, e.g. the distribution of
-  words-per-line for a word count program.
+  和 {% gh_link /flink-core/src/main/java/org/apache/flink/api/common/accumulators/DoubleCounter.java "__DoubleCounter__" %}:
+  有关使用计数器的示例，请参见下文。
+- {% gh_link /flink-core/src/main/java/org/apache/flink/api/common/accumulators/Histogram.java "__直方图__" %}:
+  离散数量的柱状直方图实现。 在内部，它只是整形到整形的映射。你可以使用它来计算值的分布，例如，单词计数程序的每行单词的分布情况。
 
-__How to use accumulators:__
+__如何使用累加器:__
```
Comment (RocMarshal) [inline docs/dev/user_defined_functions.zh.md:173]:
```suggestion
__如何使用累加器：__
```

File: docs/dev/user_defined_functions.zh.md [hunk 6]
```diff
@@ -62,32 +60,31 @@ data.filter(s -> s.startsWith("http://"));
...
+你可以写成
 
 {% highlight java %}
 class MyMapFunction extends RichMapFunction<String, Integer> {
   public Integer map(String value) { return Integer.parseInt(value); }
 };
 {% endhighlight %}
 
-and pass the function as usual to a `map` transformation:
+并将函数照常传递给map算子:
```
[the remark this answers is not in the document]
Comment (RocMarshal) [reply docs/dev/user_defined_functions.zh.md:81]: The key word `transformation` is a significant word for conceptions of Flink.

**Review by RocMarshal — CHANGES_REQUESTED** [review]
@zhuxiaoshang  Thanks for your contribution, the change looks in a good shape now, left some comments, please let me know what do you think about it. 

In addition, you can also do some optimization according to the requirements of these pages ,[Flink Translation Specifications](https://cwiki.apache.org/confluence/display/FLINK/Flink+Translation+Specifications), [贡献文档](https://flink.apache.org/zh/contributing/contribute-documentation.html) etc.
Thank you.

File: docs/dev/user_defined_functions.zh.md (hunk 1 above)
Comment (zhuxiaoshang) [reply docs/dev/user_defined_functions.zh.md:160]: These link-tags can be linked to correct pages,so i think it's unnecessary to change.

**Comment by zhuxiaoshang** [comment]
Thanks for your review @RocMarshal.I have made some changes according to your suggestions.

File: docs/dev/user_defined_functions.zh.md (hunk 1 above)
Comment (RocMarshal) [reply docs/dev/user_defined_functions.zh.md:160]: @klion26 What do you think about this?

File: docs/dev/user_defined_functions.zh.md [hunk 7]
```diff
@@ -147,95 +153,77 @@ data.map (new RichMapFunction[String, Int] {
...
-reduce, etc), four methods: `open`, `close`, `getRuntimeContext`, and
-`setRuntimeContext`. These are useful for parameterizing the function
-(see [Passing Parameters to Functions]({{ site.baseurl }}/dev/batch/index.html#passing-parameters-to-functions)),
-creating and finalizing local state, accessing broadcast variables (see
-[Broadcast Variables]({{ site.baseurl }}/dev/batch/index.html#broadcast-variables)), and for accessing runtime
-information such as accumulators and counters (see
-[Accumulators and Counters](#accumulators--counters)), and information
-on iterations (see [Iterations]({{ site.baseurl }}/dev/batch/iterations.html)).
+除了用户自定义的功能（map，reduce 等），Rich functions 还提供了四个方法：`open`、`close`、`getRuntimeContext` 和
+`setRuntimeContext`。这些对于参数化功能很有用
```
Comment (klion26) [inline docs/dev/user_defined_functions.zh.md:157]: `这些对于参数化功能很有用` 这个单纯读中文觉得有点奇怪，能否优化一下呢？

File: docs/dev/user_defined_functions.zh.md [hunk 8]
```diff
@@ -147,95 +153,77 @@ data.map (new RichMapFunction[String, Int] {
...
+(参阅 [给函数传递参数]({{ site.baseurl }}/zh/dev/batch/index.html#passing-parameters-to-functions))，
```
Comment (klion26) [inline docs/dev/user_defined_functions.zh.md:158]: 像 @RocMarshal 说的一样，现在的链接，建议使用 `{%link %}` 的形式，具体原因可以看他给出的链接

File: docs/dev/user_defined_functions.zh.md [hunk 9]
```diff
@@ -147,95 +153,77 @@ data.map (new RichMapFunction[String, Int] {
...
+创建和最终确定本地状态，访问广播变量(参阅
+[广播变量]({{ site.baseurl }}/zh/dev/batch/index.html#broadcast-variables))，以及访问运行时信息，例如累加器和计数器(参阅
+[累加器和计数器](#累加器和计数器))，以及迭代器的相关信息(参阅 [迭代器]({{ site.baseurl }}/zh/dev/batch/iterations.html))。
```
Comment (klion26) [inline docs/dev/user_defined_functions.zh.md:161]: `[累加器和计数器](#累加器和计数器)` 这里的跳转不要用中文

File: docs/dev/user_defined_functions.zh.md [hunk 10]
```diff
@@ -147,95 +153,77 @@ data.map (new RichMapFunction[String, Int] {
...
+  有关使用计数器的示例，请参见下文。
+- {% gh_link /flink-core/src/main/java/org/apache/flink/api/common/accumulators/Histogram.java "__直方图__" %}:
+  离散数量的柱状直方图实现。在内部，它只是整形到整形的映射。你可以使用它来计算值的分布，例如，单词计数程序的每行单词的分布情况。
 
-__How to use accumulators:__
+__如何使用累加器：__
 
-First you have to create an accumulator object (here a counter) in the user-defined transformation
-function where you want to use it.
+首先，你要在需要使用累加器的用户自定义的转换函数中创建一个累加器对象（此处是计数器）。
```
Comment (klion26) [inline docs/dev/user_defined_functions.zh.md:188]:
```suggestion
首先，在需要使用累加器的用户自定义函数中创建一个累加器对象（此处是计数器）。
```

File: docs/dev/user_defined_functions.zh.md [hunk 11]
```diff
@@ -147,95 +153,77 @@ data.map (new RichMapFunction[String, Int] {
...
 
 {% highlight java %}
 private IntCounter numLines = new IntCounter();
 {% endhighlight %}
 
-Second you have to register the accumulator object, typically in the ```open()``` method of the
-*rich* function. Here you also define the name.
+其次，你必须在富函数的```open()```方法中注册累加器对象。也可以在此处定义名称。
```
Comment (klion26) [inline docs/dev/user_defined_functions.zh.md:194]:
```suggestion
其次，你必须在富函数的 ```open()``` 方法中注册累加器对象。也可以在此处定义名称。
```

File: docs/dev/user_defined_functions.zh.md [hunk 12]
```diff
@@ -147,95 +153,77 @@ data.map (new RichMapFunction[String, Int] {
...
+{% gh_link /flink-java/src/main/java/org/apache/flink/api/java/operators/IterativeDataSet.java#L98 "聚合器" %}
+来计算每次迭代的统计信息，并基于此类统计信息来终止迭代。
 
-__Custom accumulators:__
+__定制累加器：__
 
-To implement your own accumulator you simply have to write your implementation of the Accumulator
-interface. Feel free to create a pull request if you think your custom accumulator should be shipped
-with Flink.
+要实现自己的累加器，你只需要实现累加器接口即可。如果你认为自定义累加器应随 Flink 一起提供，请尽管创建拉取请求。
```
Comment (klion26) [inline docs/dev/user_defined_functions.zh.md:220]: `拉取请求` 我个人建议保持 `pull request` 原文，中文不一定知道是什么意思

File: docs/dev/user_defined_functions.zh.md [hunk 13]
```diff
@@ -147,95 +153,77 @@ data.map (new RichMapFunction[String, Int] {
...
-You have the choice to implement either
+你可以选择实现
 {% gh_link /flink-core/src/main/java/org/apache/flink/api/common/accumulators/Accumulator.java "Accumulator" %}
-or {% gh_link /flink-core/src/main/java/org/apache/flink/api/common/accumulators/SimpleAccumulator.java "SimpleAccumulator" %}.
+或 {% gh_link /flink-core/src/main/java/org/apache/flink/api/common/accumulators/SimpleAccumulator.java "SimpleAccumulator" %}。
 
-```Accumulator<V,R>``` is most flexible: It defines a type ```V``` for the value to add, and a
-result type ```R``` for the final result. E.g. for a histogram, ```V``` is a number and ```R``` is
- a histogram. ```SimpleAccumulator``` is for the cases where both types are the same, e.g. for counters.
+```Accumulator<V,R>``` 最灵活: 它给需要添加的值定义了类型```V```，并定义了最终的结果类型```R```。例如，对于直方图，```V```是一个数字且```R```是一个直方图。
```
Comment (klion26) [inline docs/dev/user_defined_functions.zh.md:226]:
```suggestion
```Accumulator<V,R>``` 最灵活: 它给需要添加的值定义了类型 ```V```，并定义了最终的结果类型 ```R```。例如，对于直方图，```V``` 是一个数字且 ```R```是一个直方图。
```

**Review by klion26 — COMMENTED** [review]
@zhuxiaoshang thanks for your contribution, left some comments, please let me know what do you think, thanks.

File: docs/dev/user_defined_functions.zh.md (hunk 8 above)
Comment (zhuxiaoshang) [reply docs/dev/user_defined_functions.zh.md:158]: 可以举个例子吗？那个链接没太看明白
现有链接好像可以正确链接到对应页面。

**Comment by zhuxiaoshang** [comment]
> @zhuxiaoshang thanks for your contribution, left some comments, please let me know what do you think, thanks.

thanks for your review,i had fixed accroading to your suggestions.

File: docs/dev/user_defined_functions.zh.md [hunk 14]
```diff
@@ -147,95 +153,78 @@ data.map (new RichMapFunction[String, Int] {
...
-`setRuntimeContext`. These are useful for parameterizing the function
-(see [Passing Parameters to Functions]({{ site.baseurl }}/dev/batch/index.html#passing-parameters-to-functions)),
-creating and finalizing local state, accessing broadcast variables (see
-[Broadcast Variables]({{ site.baseurl }}/dev/batch/index.html#broadcast-variables)), and for accessing runtime
-information such as accumulators and counters (see
-[Accumulators and Counters](#accumulators--counters)), and information
-on iterations (see [Iterations]({{ site.baseurl }}/dev/batch/iterations.html)).
+除了用户自定义的功能（map，reduce 等），Rich functions 还提供了四个方法：`open`、`close`、`getRuntimeContext` 和
+`setRuntimeContext`。这些方法对于参数化函数
+(参阅 [给函数传递参数]({{ site.baseurl }}{% link zh/dev/batch/index.html#passing-parameters-to-functions %}))，
```
Comment (klion26) [inline docs/dev/user_defined_functions.zh.md:158]: 这个链接的修改是有问题的，具体可以参考下上次评论中的链接。
http://apache-flink-mailing-list-archive.1008284.n3.nabble.com/Reminder-Prefer-link-tag-in-documentation-td42362.html

需要把
[CLI]({% site.baseurl %}/ops/cli.html) 的格式换成 [CLI]({% link ops/cli.md %})。

你可以参考 https://github.com/apache/flink/pull/12798/files 这个 PR 的修改。

修改完成之后，建议执行 `./docs/docker/run.sh` 然后执行 `./build_docs.sh -p`  查看是否有错误（现在的修改执行是有错误的），这样执行没有错误之后，可以打开 `localhost:4000` 找到翻译的页面查看具体效果

**Review by klion26 — COMMENTED** [review]
@zhuxiaoshang thanks for your contribution, but the change will fail the build, I think we should fix the failure first.

Please see the inline comment for more detail. thanks

File: docs/dev/user_defined_functions.zh.md (hunk 14 above)
Comment (zhuxiaoshang) [reply docs/dev/user_defined_functions.zh.md]: @klion26 yes,i have fixed it.

Comment (klion26) [reply docs/dev/user_defined_functions.zh.md]: 现在的链接还是有问题的，如果在本地执行 `./docs/docker/run.sh` 然后执行 `./build_docs.sh -p` 是可以看到报错的
正确的应该是 `{%link dev/batch/index.zh.md %}#passing-parameters-to-functions`

File: flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/operations/utils/SortOperationFactory.java [hunk 15]
```diff
@@ -121,9 +121,9 @@ private SortQueryOperation validateAndGetChildSort(QueryOperation child, PostRes
 			previousSort = (SortQueryOperation) createSort(Collections.emptyList(), child, postResolverFactory);
```
Comment (klion26) [inline flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/operations/utils/SortOperationFactory.java:121]: could you please revert these unrelated changes?

File: flink-table/flink-table-api-java/src/test/java/org/apache/flink/table/api/Fetch.java [hunk 16]
```diff
@@ -0,0 +1,13 @@
+package org.apache.flink.table.api;
```
Comment (klion26) [inline flink-table/flink-table-api-java/src/test/java/org/apache/flink/table/api/Fetch.java:1]: please revert these changes

Comment (zhuxiaoshang) [reply flink-table/flink-table-api-java/src/test/java/org/apache/flink/table/api/Fetch.java]: sorry,i have reverted these.

File: docs/dev/user_defined_functions.zh.md [hunk 17]
```diff
@@ -62,32 +66,33 @@ data.filter(s -> s.startsWith("http://"));
 data.reduce((i1,i2) -> i1 + i2);
 {% endhighlight %}
 
+<a name="rich-functions"></a>
+
 ## Rich functions
 
-All transformations that require a user-defined function can
-instead take as argument a *rich* function. For example, instead of
+所有需要用户自定义函数的转化操作都可以将 *rich* function 作为参数。例如，代替
```
Comment (RocMarshal) [inline docs/dev/user_defined_functions.zh.md:73]:
```suggestion
所有需要用户自定义函数的转化操作都可以将 *rich* function 作为参数。例如，你可以将以下代码
```

File: docs/dev/user_defined_functions.zh.md [hunk 18]
```diff
@@ -62,32 +66,33 @@ data.filter(s -> s.startsWith("http://"));
...
 
 {% highlight java %}
 class MyMapFunction implements MapFunction<String, Integer> {
   public Integer map(String value) { return Integer.parseInt(value); }
 };
 {% endhighlight %}
 
-you can write
+你可以写成
```
Comment (RocMarshal) [inline docs/dev/user_defined_functions.zh.md:81]:
```suggestion
更改后替换为如下内容
```

File: docs/dev/user_defined_functions.zh.md [hunk 19]
```diff
@@ -23,16 +23,16 @@ specific language governing permissions and limitations
 under the License.
 -->
 
-Most operations require a user-defined function. This section lists different
-ways of how they can be specified. We also cover `Accumulators`, which can be
-used to gain insights into your Flink application.
+大多数操作都需要用户自定义函数。本节列出了实现用户自定义函数的不同方式。还会介绍 `Accumulators`（累加器），可用于深入了解你的 Flink 应用程序。
```
Comment (RocMarshal) [inline docs/dev/user_defined_functions.zh.md:26]:
```suggestion
大多数操作都需要用户自定义函数（function）。本节列出了实现用户自定义函数的不同方式。还会介绍 `Accumulators`（累加器），可用于深入了解你的 Flink 应用程序。
```

File: docs/dev/user_defined_functions.zh.md [hunk 20]
```diff
@@ -114,30 +119,31 @@ data.reduce { (i1,i2) => i1 + i2 }
...
 
-and pass the function to a `map` transformation:
+并将函数传递给 `map` transformation:
 
 {% highlight scala %}
 data.map(new MyMapFunction())
 {% endhighlight %}
 
-Rich functions can also be defined as an anonymous class:
+富函数也可以定义成匿名类:
```
Comment (RocMarshal) [inline docs/dev/user_defined_functions.zh.md:146]: We need to make some necessary decisions.
```function``` or ```函数```？
```rich function``` or ```富函数````？
We'd better keep the special nouns and keywords consistent in the translation, which  is more rigorous.
Can you tell me what you think of ?

File: docs/dev/user_defined_functions.zh.md [hunk 21]
```diff
@@ -147,95 +153,78 @@ data.map (new RichMapFunction[String, Int] {
...
-You have the choice to implement either
+你可以选择实现
 {% gh_link /flink-core/src/main/java/org/apache/flink/api/common/accumulators/Accumulator.java "Accumulator" %}
-or {% gh_link /flink-core/src/main/java/org/apache/flink/api/common/accumulators/SimpleAccumulator.java "SimpleAccumulator" %}.
+或 {% gh_link /flink-core/src/main/java/org/apache/flink/api/common/accumulators/SimpleAccumulator.java "SimpleAccumulator" %}。
 
-```Accumulator<V,R>``` is most flexible: It defines a type ```V``` for the value to add, and a
-result type ```R``` for the final result. E.g. for a histogram, ```V``` is a number and ```R``` is
- a histogram. ```SimpleAccumulator``` is for the cases where both types are the same, e.g. for counters.
+```Accumulator<V,R>``` 最灵活: 它给需要添加的值定义了类型 ```V```，并定义了最终的结果类型 ```R```。例如，对于直方图，```V``` 是一个数字且 ```R``` 是一个直方图。
```
Comment (RocMarshal) [inline docs/dev/user_defined_functions.zh.md:227]:
```suggestion
```Accumulator<V,R>``` 的实现十分灵活: 它定义了将要添加的值类型 ```V```，并定义了最终的结果类型 ```R```。例如，对于直方图，```V``` 是一个数字且 ```R``` 是一个直方图。
```

**Review by RocMarshal — COMMENTED** [review]
Thanks for your effort @zhuxiaoshang . The overall translation looks good. I only had a few minor comments.

File: docs/dev/user_defined_functions.zh.md (hunk 20 above)
Comment (zhuxiaoshang) [reply docs/dev/user_defined_functions.zh.md:146]: yes，it's should be resvered as `function`

File: docs/dev/user_defined_functions.zh.md [hunk 22]
```diff
@@ -147,95 +153,78 @@ data.map (new RichMapFunction[String, Int] {
...
-`setRuntimeContext`. These are useful for parameterizing the function
-(see [Passing Parameters to Functions]({{ site.baseurl }}/dev/batch/index.html#passing-parameters-to-functions)),
-creating and finalizing local state, accessing broadcast variables (see
-[Broadcast Variables]({{ site.baseurl }}/dev/batch/index.html#broadcast-variables)), and for accessing runtime
-information such as accumulators and counters (see
-[Accumulators and Counters](#accumulators--counters)), and information
-on iterations (see [Iterations]({{ site.baseurl }}/dev/batch/iterations.html)).
+除了用户自定义的 function（map，reduce 等），Rich functions 还提供了四个方法：`open`、`close`、`getRuntimeContext` 和
+`setRuntimeContext`。这些方法对于参数化 function
+(参阅 [给 function 传递参数]({% link dev/batch/index.zh.md %}#passing-parameters-to-functions))，
```
Comment (RocMarshal) [inline docs/dev/user_defined_functions.zh.md:158]:
```suggestion
（参阅 [给 function 传递参数]({% link dev/batch/index.zh.md %}#passing-parameters-to-functions)），
```

File: docs/dev/user_defined_functions.zh.md [hunk 23]
```diff
@@ -147,95 +153,78 @@ data.map (new RichMapFunction[String, Int] {
...
+创建和最终确定本地状态，访问广播变量(参阅
```
Comment (RocMarshal) [inline docs/dev/user_defined_functions.zh.md:159]:
```suggestion
创建和最终确定本地状态，访问广播变量（参阅
```

File: docs/dev/user_defined_functions.zh.md [hunk 24]
```diff
@@ -147,95 +153,78 @@ data.map (new RichMapFunction[String, Int] {
...
+[广播变量]({% link dev/batch/index.zh.md %}#broadcast-variables ))，以及访问运行时信息，例如累加器和计数器(参阅
```
Comment (RocMarshal) [inline docs/dev/user_defined_functions.zh.md:160]:
```suggestion
[广播变量]({% link dev/batch/index.zh.md %}#broadcast-variables )），以及访问运行时信息，例如累加器和计数器（参阅
```

File: docs/dev/user_defined_functions.zh.md [hunk 25]
```diff
@@ -147,95 +153,78 @@ data.map (new RichMapFunction[String, Int] {
...
+[累加器和计数器](#accumulators--counters))，以及迭代器的相关信息(参阅 [迭代器]({% link dev/batch/iterations.zh.md %}))
```
Comment (RocMarshal) [inline docs/dev/user_defined_functions.zh.md:161]:
```suggestion
[累加器和计数器](#accumulators--counters)），以及迭代器的相关信息（参阅 [迭代器]({% link dev/batch/iterations.zh.md %})）
```

File: docs/dev/user_defined_functions.zh.md [hunk 26]
```diff
@@ -147,95 +153,78 @@ data.map (new RichMapFunction[String, Int] {
...
-```Accumulator.add(V value)``` method. At the end of the job Flink will sum up (merge) all partial
-results and send the result to the client. Accumulators are useful during debugging or if you
-quickly want to find out more about your data.
+累加器是具有**加法运算**和**最终累加结果**的一种简单结构，可在作业结束后使用。
 
-Flink currently has the following **built-in accumulators**. Each of them implements the
-{% gh_link /flink-core/src/main/java/org/apache/flink/api/common/accumulators/Accumulator.java "Accumulator" %}
-interface.
+最简单的累加器就是**计数器**: 你可以使用
+```Accumulator.add(V value)``` 方法将其递增。在作业结束时，Flink 会汇总(合并)所有部分的结果并将其发送给客户端。
```
Comment (RocMarshal) [inline docs/dev/user_defined_functions.zh.md:173]:
```suggestion
```Accumulator.add(V value)``` 方法将其递增。在作业结束时，Flink 会汇总（合并）所有部分的结果并将其发送给客户端。
```

**Review by RocMarshal — COMMENTED** [review]
@zhuxiaoshang  thanks for your contribution, left some comments based on the item `5. 中文标点符号` of [Flink+Translation+Specifications](https://cwiki.apache.org/confluence/display/FLINK/Flink+Translation+Specifications) please let me know what do you think ,thanks.

**Comment by zhuxiaoshang** [comment]
> @zhuxiaoshang thanks for your contribution, left some comments based on the item `5. 中文标点符号` of [Flink+Translation+Specifications](https://cwiki.apache.org/confluence/display/FLINK/Flink+Translation+Specifications) please let me know what do you think ,thanks.

yes,you are right,brackets should be chinese form.

**Review by RocMarshal — APPROVED** [review]
Thanks for the contribution. LGTM. Now ping @klion26

File: docs/dev/user_defined_functions.zh.md [hunk 27]
```diff
@@ -147,95 +153,78 @@ data.map (new RichMapFunction[String, Int] {
...
-creating and finalizing local state, accessing broadcast variables (see
-[Broadcast Variables]({{ site.baseurl }}/dev/batch/index.html#broadcast-variables)), and for accessing runtime
-information such as accumulators and counters (see
-[Accumulators and Counters](#accumulators--counters)), and information
-on iterations (see [Iterations]({{ site.baseurl }}/dev/batch/iterations.html)).
+除了用户自定义的 function（map，reduce 等），Rich functions 还提供了四个方法：`open`、`close`、`getRuntimeContext` 和
+`setRuntimeContext`。这些方法对于参数化 function
+（参阅 [给 function 传递参数]({% link dev/batch/index.zh.md %}#passing-parameters-to-functions)），
+创建和最终确定本地状态，访问广播变量（参阅
+[广播变量]({% link dev/batch/index.zh.md %}#broadcast-variables )），以及访问运行时信息，例如累加器和计数器（参阅
```
Comment (klion26) [inline docs/dev/user_defined_functions.zh.md:160]: @zhuxiaoshang @RocMarshal 这个地方的链接我怀疑是有问题的，我看了下，这个地方是 FLINK-17009 引入的，但是没有找到更早的历史。
怀疑的原因是，这个文档讲的是 datastream 的 udfs，但是这里的链接是 batch 相关的。
不过这个和整体的翻译无关，能否记录并查找一下这个地方是否真的有问题，正确的链接应该是什么？如果有问题的话，可以再提一个 PR 来进行修改

**Review by klion26 — APPROVED** [review]
@zhuxiaoshang thanks for the contribution and @RocMarshal for the review. the change LGTM
ping @wuchong @rmetzger

**Comment by rmetzger** [comment]
Thanks for the review! Merging ...
